### PR TITLE
#22 Introduce ability for admin to check trips (keeping manage expenses, …

### LINKED ISSUE #22: As an administrator I want to track trip progress

### PR DESCRIPTION
**Pull request by MxmIv**
Resolves #22

Introduce ability for admin to check trips (keeping manage expenses, but removing the edit time ability)

- Introducing the page visibility for admins and drivers 
- Admins cannot see the edit time button 
- Drivers still have the ability to edit time 
- Added/updated testing: 
->  TripItemViewModelTests: This class deals with individual trip items and the user’s ability to edit and save changes to a trip.
->  TripViewModelTests: This class represents a higher-level view of all the trips and handles loading the trip data into TripItemViewModel instances.

Testing: 
- Tested the whole flow locally 

**Driver login:** 
Screenshot 2024-08-23 at 23 06 20
Screenshot 2024-08-24 at 23 11 55

**Admin login:** 
Screenshot 2024-08-24 at 23 12 03

Screenshot 2024-08-24 at 23 12 37

### CODE REVIEW

File: HaulageApplication/HaulageApp/Services/UserService.cs [hunk 1]
```diff
@@ -1,9 +1,24 @@
...
+        public UserService(HaulageDbContext context)
+        {
+            _context = context;
+        }
+
+        public bool IsDriver()
+        {
+            var currentUserId = GetCurrentUserId();
+            var user = _context.user.FirstOrDefault(u => u.Id == currentUserId);
+            return user != null && user.Role == 2;
```
Comment (tiphainetiphaine) [inline HaulageApplication/HaulageApp/Services/UserService.cs:19]: Should this be done properly and a check to Role Id where Role type is "driver" be carried out? I.e. a call/join to the Role table.

File: HaulageApplication/HaulageApp/Views/TripPage.xaml.cs [hunk 2]
```diff
@@ -4,10 +4,22 @@ namespace HaulageApp.Views
...
+            _viewModel = viewModel;
+            BindingContext = _viewModel;
+        }
+
+        protected override async void OnAppearing()
+        {
+            base.OnAppearing();
+            if (_viewModel != null)
+            {
+                await _viewModel.LoadDataAsync();
```
Comment (tiphainetiphaine) [inline HaulageApplication/HaulageApp/Views/TripPage.xaml.cs:21]: Why is this here rather than in the constructor? 
Do you want to force a reload on the view appearing?

File: HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs [hunk 3]
```diff
@@ -11,45 +11,40 @@ namespace HaulageApp.ViewModels
...
 
-        public TripItemViewModel(Trip trip, HaulageDbContext context)
+        public TripItemViewModel(Trip trip, HaulageDbContext context, bool canEdit)
         {
             _trip = trip;
             _context = context;
-
+            _canEdit = canEdit;
+            
             _tempStartTimeString = _trip.StartTime.ToString("yyyy-MM-dd HH:mm", CultureInfo.InvariantCulture);
```
Comment (tiphainetiphaine) [inline HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs:30]: NIT: It may be better to avoid parsing in the ViewModel directly, to decouple string formatting from the view model to make it cleaner and easier to manage.

File: HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs [hunk 4]
```diff
@@ -94,14 +89,14 @@ public string EndTimeString
...
-            get => _isEditing;
+            get => _isEditing && _canEdit;
             set
             {
                 if (_isEditing != value)
                 {
                     _isEditing = value;
                     OnPropertyChanged(nameof(IsEditing));
-                    OnPropertyChanged(nameof(EditSaveButtonText)); // Notify change here to update button text
+                    OnPropertyChanged(nameof(EditSaveButtonText));
```
Comment (tiphainetiphaine) [inline HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs:107]: Is this property interdependent with IsEditing? If so, do you need a call to both (OnPropertyChanged)? If not, you can reduce the redundant call.

File: HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs
Comment (tiphainetiphaine) [inline HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs:1]: I noticed this but it doesn't relate to this CR so I can't comment on it directly, but you create a new waypoint collection every time it's accessed:

`public ObservableCollection<Waypoint> Waypoints => new ObservableCollection<Waypoint>(_trip.Waypoints);`

Is there a reason this is not previously initialised and then updated/bound?

File: HaulageApplication/HaulageApp/ViewModels/TripViewModel.cs [hunk 5]
```diff
@@ -20,26 +22,29 @@ public ObservableCollection<TripItemViewModel> TripGroups
...
         {
             try
             {
+                bool canEdit = _userService.IsDriver();
+
                 var trips = await _context.trip
                     .Include(t => t.Waypoints)
                     .ToListAsync();
 
                 TripGroups = new ObservableCollection<TripItemViewModel>(
```
Comment (tiphainetiphaine) [inline HaulageApplication/HaulageApp/ViewModels/TripViewModel.cs:41]: Why is a "new" observable collection being created?
Have you considered updating the existing collection?

File: HaulageApplication/HaulageAppTests/TripViewModelTests.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            _userServiceMock = new Mock<IUserService>();
+            _options = new DbContextOptionsBuilder<HaulageDbContext>()
+                .UseInMemoryDatabase(databaseName: Guid.NewGuid().ToString())
+                .Options;
+        }
+
+        private void ClearDatabase()
+        {
+            using (var context = new HaulageDbContext(_options))
```
Comment (tiphainetiphaine) [inline HaulageApplication/HaulageAppTests/TripViewModelTests.cs:26]: Since this is being created multiple times, consider creating a shared context method.

File: HaulageApplication/HaulageAppTests/TripViewModelTests.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        [Fact]
+        public async Task TripViewModel_LoadsDataAsync_ForDriver()
+        {
+            ClearDatabase();
+
+            using (var context = new HaulageDbContext(_options))
+            {
+                context.trip.AddRange(
```
Comment (tiphainetiphaine) [inline HaulageApplication/HaulageAppTests/TripViewModelTests.cs:40]: Was there a reason you chose not to add these to the mock db file, preferring to inline? 
Regardless, maybe this could be set up once, since I see this is set up multiple times in this file

File: HaulageApplication/HaulageAppTests/TripViewModelTests.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+            Assert.NotNull(viewModel.TripGroups);
+            Assert.Equal(2, viewModel.TripGroups.Count);
+            Assert.All(viewModel.TripGroups, t => Assert.True(t.IsEditButtonVisible));
+        }
+
+        [Fact]
+        public async Task TripViewModel_LoadsDataAsync_ForAdmin()
+        {
+            ClearDatabase();
```
Comment (tiphainetiphaine) [inline HaulageApplication/HaulageAppTests/TripViewModelTests.cs:61]: NIT if you are going to do this, then you can add it as [setup] and [teardown] methods

File: HaulageApplication/HaulageApp/Views/TripPage.xaml.cs (hunk 2 above)
Comment (tiphainetiphaine) [reply HaulageApplication/HaulageApp/Views/TripPage.xaml.cs:21]: I've noticed this is async so it can't be in the constructor. But I guess the question remains, do you want to (re) load the data on each appear?

File: HaulageApplication/HaulageApp/Services/UserService.cs (hunk 1 above)
Comment (MxmIv) [reply HaulageApplication/HaulageApp/Services/UserService.cs:19]: The role check is done properly by directly using the role ID in the user table. This approach avoids unnecessary joins, keeping the code simple and efficient. Since role IDs are stable and clearly defined (1 for customer, 2 for driver, 3 for admin), there's no ambiguity or risk. This method ensures better performance, adheres to the KISS principle, and fits our current project scope without introducing unnecessary complexity. If future needs change, we can revisit and refactor accordingly.

File: HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs (hunk 4 above)
Comment (MxmIv) [reply HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs:107]: `EditSaveButtonText` depends on `IsEditing`, both calls to OnPropertyChanged are needed and should be kept. This is not redundant, it's ensuring that all dependent properties are updated when `IsEditing` changes.

File: HaulageApplication/HaulageApp/Views/TripPage.xaml.cs (hunk 2 above)
Comment (MxmIv) [reply HaulageApplication/HaulageApp/Views/TripPage.xaml.cs:21]: This was made on purpose due to these scenarios: 
- OnAppearing() method is called every time the page appears in the user interface, not just when it is first created. This makes it ideal for reloading data to ensure that the UI reflects the most current state. For example, if the user navigates away from the page and then comes back, you might want to refresh the data to reflect any changes (trip updates) that occurred while the user was on another screen.
-  OnAppearing() is particularly important when different users might sign in/out in the same session  (while the app is running) to make sure that the current user's data is displayed properly

File: HaulageApplication/HaulageApp/ViewModels/TripViewModel.cs (hunk 5 above)
Comment (MxmIv) [reply HaulageApplication/HaulageApp/ViewModels/TripViewModel.cs:41]: Yes, when i was not creating new collection the existing collection the app was having weird UX flickering on update, additionally it was working a bit slower. I can create a ticket to add it to the backlog to revisit 
https://github.com/user-attachments/assets/fa4a9d81-99c0-4f43-84fc-8a31bc8a6d53

File: HaulageApplication/HaulageAppTests/TripViewModelTests.cs (hunk 6 above)
Comment (MxmIv) [reply HaulageApplication/HaulageAppTests/TripViewModelTests.cs:26]: done

File: HaulageApplication/HaulageAppTests/TripViewModelTests.cs (hunk 7 above)
Comment (MxmIv) [reply HaulageApplication/HaulageAppTests/TripViewModelTests.cs:40]: No, i have addressed now and moved this  to `MockDb` the adhere to DRY (Don't Repeat Yourself) principle

File: HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs
Comment (MxmIv) [reply HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs:1]: there is no particular reason. I will need to come back to this to give you a clearer answer at this point.

**Review by tiphainetiphaine — APPROVED** [review]
